win32ss/printing/base/winspool/printers.c

```diff
@@ -384,10 +384,100 @@ EndPagePrinter(HANDLE hPrinter)
+
+    TRACE("EnumPrintersA(%lu, %s, %lu, %p, %lu, %p, %p)\n", Flags, pName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
+
+    if (pName)
+    {
+        // Convert pName to a Unicode string pwszName.
+        cch = strlen(pName);
+
+        pwszName = HeapAlloc(hProcessHeap, 0, (cch + 1) * sizeof(WCHAR));
+        if (!pwszName)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            ERR("HeapAlloc failed!\n");
+            goto Cleanup;
+        }
+
+        MultiByteToWideChar(CP_ACP, 0, pName, -1, pwszName, cch + 1);
+    }
+
+    bReturnValue = EnumPrintersW(Flags, pwszName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
+
+    TRACE("*pcReturned is '%d' and bReturnValue is '%d' and GetLastError is '%ld'.\n", *pcReturned, bReturnValue, GetLastError());
+
+    /* We are mapping two different pointers to the pPrinterEnum pointer here so that */
+    /* we can do in-place conversion. We read the Unicode response from the EnumPrintersW and */
+    /* then we write back the ASCII conversion into the same buffer for our EnumPrintersA output */
+
+    /* mapping to pPrinterEnum for Unicode (w) characters */
+    ppi4w = (PPRINTER_INFO_4W)pPrinterEnum;
+    /* mapping to pPrinterEnum for ASCII (a) characters */
+    ppi4a = (PPRINTER_INFO_4A)pPrinterEnum;
+
+    for (i = 0; i < *pcReturned; i++)
+    {
+        if (ppi4w[i].pPrinterName)
+        {
+            // Convert Unicode pPrinterName to a ASCII string pszPrinterName.
+            cch = wcslen(ppi4w[i].pPrinterName);
+
+            pszPrinterName = HeapAlloc(hProcessHeap, 0, (cch + 1) * sizeof(CHAR));
+            if (!pszPrinterName)
+            {
+                SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+                ERR("HeapAlloc failed!\n");
+                goto Cleanup;
+            }
+
+            WideCharToMultiByte(CP_ACP, 0, ppi4w[i].pPrinterName, -1, pszPrinterName, cch + 1, NULL, NULL);
+            strcpy(ppi4a[i].pPrinterName, pszPrinterName);
+
+            HeapFree(hProcessHeap, 0, pszPrinterName);
+        }
+
+        if (ppi4w[i].pServerName)
+        {
+            // Convert Unicode pServerName to a ASCII string pszServerName.
+            cch = wcslen(ppi4w[i].pServerName);
+
+            pszServerName = HeapAlloc(hProcessHeap, 0, (cch + 1) * sizeof(CHAR));
+            if (!pszServerName)
+            {
+                SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+                ERR("HeapAlloc failed!\n");
+                goto Cleanup;
+            }
+
+            WideCharToMultiByte(CP_ACP, 0, ppi4w[i].pServerName, -1, pszServerName, cch + 1, NULL, NULL);
+            strcpy(ppi4a[i].pServerName, pszServerName);
+
+            HeapFree(hProcessHeap, 0, pszServerName);
+        }
+    }
+
+Cleanup:
+    if (pwszName)
+        HeapFree(hProcessHeap, 0, pwszName);
+
+    if (pszPrinterName)
+        HeapFree(hProcessHeap, 0, pszPrinterName);
+
+    if (pszServerName)
+        HeapFree(hProcessHeap, 0, pszServerName);
```

Comment:
These shouldn't be here for two reasons:
1. There's no code path through which we can get here in which either of these strings was allocated but not freed. They're used in their respective local blocks only. So the free here is unnecessary.
2. Since you don't null the pointers above, this will result in a double free.

I'd be okay with moving the declarations for these variables into their `if` blocks to make their usage clearer; however most of our code keeps all variables at the top of the function as a style choice. Your call.

Reply:
Fixed.